server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -159,10 +162,10 @@ public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStore
 
     private Mono<byte[]> readBytesInDefaultBucket(BucketName bucketName, BlobId blobId) {
         return readFromCache(blobId)
-            .switchIfEmpty(readBytesFromBackend(bucketName, blobId)
+            .switchIfEmpty(readBytesFromBackend(bucketName, blobId))
                 .filter(this::isAbleToCache)
                 .doOnNext(any -> metricRetrieveMissCount.increment())
-                .flatMap(bytes -> saveInCache(blobId, bytes).then(Mono.just(bytes))));
+                .flatMap(bytes -> saveInCache(blobId, bytes).then(Mono.just(bytes)));
```

Comment:
The indent don't reflect the pipeline being executed. There's a red light here: I believe you cache things after cache hits.

Please write extra tests (we don't have enough if you were able to write this).

Reply:
A logic here was wrong also.

Reply:
Which test proves it wrong?

Also you introduced new undesired behaviours, wich we should actually be testing...

Reply:
How about this:
```
return readFromCache(blobId).switchIfEmpty(
            readBytesFromBackend(bucketName, blobId)
                .flatMap(bytes -> {
                    if (isAbleToCache(bytes)) {
                        return saveInCache(blobId, bytes)
                            .doOnSuccess(any -> metricRetrieveMissCount.increment())
                            .then(Mono.just(bytes));
                    }
                    return Mono.just(bytes);
                }));
```